### ISSUE TITLE
Provide JSON-formatted output

### ISSUE THREAD

**Issue by rlipscombe** [electricimp/imp-central-impt#16]
So that I can parse it with jq.

Yeah, I get that the API already outputs JSON, but this way I only have to learn one tool, and I don't have to worry about pagination, login, etc.

**Comment by rlipscombe** [comment]
tbh, wherever it outputs something that I might want to use in a later command, but for now, just the `impt X list` commands:

```
$ impt device list
Device list (1 items):
Device: 
  id:            52414c0000000001
  mac_address:   52:41:4c:00:00:01
  agent_id:      oLmfd5Y9uPSC
  device_online: true
  Device Group: 
    id:   47762fb4-531b-b517-5763-c3be0203a5d5
    type: development
    name: Development
```

It'd be nice if `impt device list --json` worked. Cos then I could `impt device list --json | jq '.[0].id'`

**Comment by ppetrosh** [comment]
That would be nice to have, agree. Thanks.

**Comment by alexey-nbl** [comment]
An idea is to introduce new group of commands - "impt setup" - with commands which control the tool behavior/output.
One of the commands, for example, would switch on/off "script" mode - a mode where meaningful output is formatted as json, other output and colors are off, etc.

**Comment by rlipscombe** [comment]
> An idea is to introduce new group of commands - "impt setup" - with commands which control the tool behavior/output.

*This is a bad idea.*

Whether or not I want JSON output is not a global setting -- if I'm running impt at the command line, I probably _don't_ want JSON. If I'm running impt in a script, then I _do_ want JSON. I shouldn't need to know whether a script is using impt in order to know whether to set the global setting. Further, if the script sets the global setting itself, how can I be sure that it's set back to whatever (unknown) value it had previously?

Just provide a `--format json` switch.

**Comment by alexey-nbl** [comment]
If I understand correctly, this is again only because you want to work in CLI with impt in parallel with the scripts working with impt. Right?

From another side, it might seem strange to add --json option, --no-color options, etc. into almost all commands...

Somebody, who use impt in a more usual way - with one impCentral - may ask: why I cannot setup the tool behavior/output once but have to repeat the options in every command?

Ok. Need to think more.

**Comment by rlipscombe** [comment]
> If I understand correctly, this is again only because you want to work in CLI with impt in parallel with the scripts working with impt. Right?

Yes. Sometimes I just want to use impt; sometimes I want to process the output from impt. I _could_ talk directly to the API with curl, but impt handles all of the paging, etc. (it does, right?), which makes it easier.

**Comment by rlipscombe** [comment]
> From another side, it might seem strange to add --json option, --no-color options, etc. into almost all commands...

I disagree. If I'm scripting using impt, I'm going to want to parse the output. _Ideally_, I'd be able to ask for JSON output.

The `--no-color` option isn't a thing -- it's implied by having stdout not be a TTY.

> Somebody, who use impt in a more usual way - with one impCentral

Not relevant. Script vs. command-line is orthogonal to how many impCentral instances I want to use it with.

> may ask: why I cannot setup the tool behavior/output once but have to repeat the options in every command?

Trust me; they won't. People who use the tool on the command line will generally never use the JSON output, so they won't need to specify it. People who script it will only do it once, and won't mind.

**Comment by alexey-nbl** [comment]
> Script vs. command-line is orthogonal to how many impCentral instances I want to use it with.
If there is one impCentral instance, it is unrealistic somebody works from CLI while a script is working at the same time in parallel.
In this case a setup inside the script does not influence CLI behavior.
So, the script may have something like "impt setup script --on" at the beginning and "impt setup script --off" at the end of the script.
Where "script" mode leaves only meaningful output, in json format, w/o colors, etc.
Why this would not work? Could you pls explain that one more time for me :)

**Comment by rlipscombe** [comment]
> If there is one impCentral instance, it is unrealistic somebody works from CLI while a script is working at the same time in parallel.

No it's not. Maybe I've got a build process that pushes some squirrel to one of my models, and while that's running, I'm doing some CLI stuff from another window.

> So, the script may have something like "impt setup script --on" at the beginning and "impt setup script --off" at the end of the script.

And if that script fails, and the `impt setup script --off` never runs, then my CLI is left in script mode, which is bad.

**Comment by alexey-nbl** [comment]
> No it's not. Maybe I've got a build process that pushes some squirrel to one of my models, and while that's running, I'm doing some CLI stuff from another window.

Ok. If you know that the second does not impact the first, that's good.

> And if that script fails, and the impt setup script --off never runs, then my CLI is left in script mode, which is bad.

If it's not a crash, the script could process the fails correctly ;)
But does not matter.

Finally, we've got all the points. Thanks for clarification and patience :)

We will add an output format option into every impt command.

It will be something like "--output \<format>" where \<format>:
"json" - minimal output, in json format, w/o colours
"nocolor" - standard output w/o colours
"debug" - additional debug info displayed
More values maybe added in the future to adjust the output for other users/customers wishes/complains.
It will be possible to specify several (non-contradict) \<format> values in one command.

Current "--debug" option will be replaced by "--output debug"

**Comment by alexey-nbl** [comment]
Is it OK for everyone?

**Comment by alexey-nbl** [comment]
We see `--output` as an option to control what is generated in the output. Including, but not only, a format of the output.
`debug` additionally displays all impCentral requests/responses during the command execution.

BTW, if a command fails, we are going to output some predefined json with an error, eg. {"error": "<err_msg>"} The exact format will be defined later.
Is it OK?
